# valid strand symbols
STRAND_SYMBOLS = ["+", "-", None]

# literals used throughout the package
CDS = "cds"
CONTIG = "contig"
EXON = "exon"
GENE = "gene"
PROTEIN = "protein"
TRANSCRIPT = "transcript"

# default species/release
DEFAULT_SPECIES = "homo_sapiens"
DEFAULT_REFERENCE = "GRCh38"
DEFAULT_RELEASE = 100
